core/client/src/main/java/alluxio/client/block/UfsBlockInStream.java

```diff
@@ -0,0 +1,244 @@
+  private final FileSystemContext mContext;
+  /** {@link UfsBlockReader} for this instance. */
+  private UfsBlockReader mReader;
+
+  /**
+   * Creates a UFS block in stream.
+   *
+   * @param context the file system context
+   * @param ufsPath the UFS path
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param blockStart the start position of the block in the UFS file
+   * @param workerNetAddress the worker network address
+   * @param options the in stream options
+   * @return the {@link UfsBlockInStream} instance
+   * @throws IOException if it fails to create {@link UfsBlockInStream}
+   */
+  public static UfsBlockInStream createUfsBlockInStream(FileSystemContext context, String ufsPath,
+      long blockId, long blockSize, long blockStart, WorkerNetAddress workerNetAddress,
+      InStreamOptions options) throws IOException {
+    Closer closer = Closer.create();
+    try {
+      BlockWorkerClient blockWorkerClient =
+          closer.register(context.createBlockWorkerClient(workerNetAddress));
+      LockBlockResult result;
+      LockBlockOptions lockBlockOptions =
+          LockBlockOptions.defaults().setUfsPath(ufsPath).setOffset(blockStart)
+              .setBlockSize(blockSize).setMaxUfsReadConcurrency(options.getMaxUfsReadConcurrency());
+
+      long timeout = System.currentTimeMillis() + Configuration
+          .getLong(PropertyKey.USER_UFS_BLOCK_OPEN_TIMEOUT_MS);
+      while (true) {
+        try {
+          result = blockWorkerClient.lockBlock(blockId, lockBlockOptions);
+          break;
+        } catch (UfsBlockAccessTokenUnavailableException e) {
+          if (System.currentTimeMillis() >= timeout) {
+            throw e;
+          }
+          LOG.debug(
+              "Failed to acquire a UFS read token because of concurrency for block {} in file {}",
+              blockId, blockId);
+          CommonUtils.sleepMs(UFS_BLOCK_OPEN_RETRY_INTERVAL_MS);
+        }
+      }
+      if (result.getLockId() >= 0) {
+        // The block exists in the Alluxio now.
+        blockWorkerClient.unlockBlock(blockId);
+        closer.close();
+        return null;
+      }
+      return new UfsBlockInStream(context, blockId, blockSize, blockWorkerClient, options);
+    } catch (AlluxioException e) {
+      closer.close();
+      throw new IOException(e);
+    } catch (IOException e) {
+      closer.close();
+      throw e;
+    }
+  }
+
+  /**
+   * Creates a new UFS block input stream.
+   *
+   * @param context the file system context
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param blockWorkerClient the block worker client
+   * @param options the instream options
+   * @throws IOException if the block is not available on the remote worker
+   */
+  private UfsBlockInStream(FileSystemContext context, long blockId, long blockSize,
+      BlockWorkerClient blockWorkerClient, InStreamOptions options) throws IOException {
```

Comment:
where do we possibly throw `IOException`? and do we need to try catch and call `mCloser` here in catch?
ideally, constructor should not throw exception. Since you have a factory method, could we move the operation that may fail there?